### PR TITLE
Sprint 30

### PR DESCRIPTION
**Pull request by ericsorenson**
# Sprint 30

## [317 Court User: Ability to make procedural notes](https://github.com/flexion/ef-cms/issues/317)
image

## [3491 Clean-Up: Case Title](https://github.com/flexion/ef-cms/issues/3491)
image

## [3897 Petitions Clerk: Issuing Orders](https://github.com/flexion/ef-cms/issues/3897)
image

## [1900 Court User: Editing Trial Session Details](https://github.com/flexion/ef-cms/issues/1900)
image

## [3290 Petitions/Docket Clerks: Flatbed and two-sided scanning (In Progress)](https://github.com/flexion/ef-cms/issues/3290)
image

### CODE REVIEW

**Comment by waldoj** [comment]
### PR Review Checklist

- [ ] begin by deleting any existing copy of the Git repository, running `docker system prune` to remove any cached artifacts from any prior code review
- [ ] clone a fresh copy of the Git repository (`git clone https://github.com/ustaxcourt/ef-cms.git`), fetch the pull request and name it for the sprint (e.g., `git fetch origin pull/204/head:sprint-43`), and then switch to that branch (e.g. `git checkout sprint-43`)
- [ ] review the pull request itself, to get oriented
	- [ ] read the description of the pull request, which should summarize the changes made
	- [ ] read through every task on the Scrum board that's encompassed by this pull request
	- [ ] read the description of the commits that comprise the pull request
- [ ] stand up the site locally, with `./docker-run.sh`
	- [ ] test all functionality in all major browsers, emphasizing the functionality that this pull request addresses
		- [ ] for internal Court functionality, perform the most thorough testing in Chrome, though also test in Edge and Firefox
		- [ ] for public-facing functionality, test in browsers consistent with [public browser use data](https://analytics.usa.gov/), including Internet Explorer 11
		- [ ] test in Mobile Safari and Mobile Chrome, with the caveat that not all internal Court functionality will be necessary on these platforms
	- [ ] use an automated audit tool for code quality and practices (recommended: [Chrome DevTools](https://developers.google.com/web/tools/chrome-devtools/), aka [Lighthouse](https://chrome.google.com/webstore/detail/lighthouse/blipmdconlkpinefehnmjammfjpmpbjk))
		- [ ] look at efficiency of page loads, asset sizes, HTTP connection management, etc.
	- [ ] review for accessibility
		- [ ] use an automated audit tool, such as Chrome Audit or aXe
		- [ ] navigate site only with the keyboard
		- [ ] use VoiceOver or Narrator to navigate the site with audio only, with the display turned off
		- [ ] manually test anything that pa11y cannot test automatically (e.g., contrast of text over images)
- [ ] [review static code analysis results](https://sonarcloud.io/organizations/flexion-github/projects) in the vendor’s SonarCloud environment
- [ ] examine OWASP ZAP output in `docs/`, to ensure that any errors are known to be false positives or have been previously declared to be acceptable
- [ ] skim all new code, in the context of existing code, [looking for problems](#what-we-look-for) (knowing that the vast majority of new code will be covered by tests)
- [ ] review all tests
	- [ ] look at code coverage of tests in [the vendor's SonarCloud environment](https://sonarcloud.io/organizations/flexion-github/projects)
	- [ ] methodically review all new tests for correctness, quality of naming
- [ ] determine what code isn’t tested, review that rigorously
- [ ] review documentation to ensure that it matches changes
- [ ] provide comments on the pull request on GitHub, as necessary
	- [ ] for comments that are specific to a particular line of code, comment on those specific lines
	- [ ] for comments that are more general, attach the comment to a random line in `README.md` (as opposed to commenting on the pull request itself), to be able to use GitHub's ability to thread discussions on those comments
- [ ] for each feature-level bug (i.e., it’s working as designed, but designed wrong), open a new issue and put it in the backlog

**Comment by waldoj** [comment]
Because GitHub lacks effective pull-request-level feedback mechanisms that must be addressed before a pull request can be merged, I'll be using `README.md` as the place to attach PR-level comments.

File: shared/src/business/useCases/trialSessions/deleteTrialSessionInteractor.test.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+    await expect(
+      deleteTrialSessionInteractor({
+        applicationContext,
+        trialSessionId: 'c54ba5a9-b37b-479d-9201-067ec6e335bb',
+      }),
+    ).rejects.toThrow('trial session not found');
+  });
+
+  it('throws error when trial session start date is past now', async () => {
```
Comment (waldoj) [inline shared/src/business/useCases/trialSessions/deleteTrialSessionInteractor.test.js]: I'm confused by this description and what's tested. You can't delete a trial session that's scheduled for 2050? How is that after the start date? Either this test needs a rethinking or I'm missing something.

File: shared/src/business/useCases/trialSessions/deleteTrialSessionInteractor.test.js [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    await expect(
+      deleteTrialSessionInteractor({
+        applicationContext,
+        trialSessionId: 'c54ba5a9-b37b-479d-9201-067ec6e335bb',
+      }),
+    ).rejects.toThrow('Trial session cannot be updated after its start date');
+  });
+
+  it('throws error when trial session is calendared', async () => {
```
Comment (waldoj) [inline shared/src/business/useCases/trialSessions/deleteTrialSessionInteractor.test.js]:
```suggestion
  it('throws error if trial session is calendared', async () => {
```
I think this is a bit clearer.

File: shared/src/persistence/dynamsoft/getScannerInterface.test.js [hunk 3]
```diff
@@ -178,6 +182,53 @@ describe('getScannerInterface', () => {
...
+    expect(DWObject.IfDuplexEnabled).toEqual(true);
+
+    await scannerAPI.startScanSession({
+      applicationContext,
+      scanMode: SCAN_MODES.FLATBED,
+    });
+    expect(DWObject.IfDuplexEnabled).toEqual(false);
+  });
+
+  it('can enable flatbed scanning by calling startScanSession with scanMode set to `FLATBED`', async () => {
```
Comment (waldoj) [inline shared/src/persistence/dynamsoft/getScannerInterface.test.js]: Why is the feeder involved in this test?

File: web-client/src/presenter/computeds/formattedTrialSessionDetails.test.js [hunk 4]
```diff
@@ -52,6 +55,70 @@ describe('formattedTrialSessionDetails', () => {
...
+        },
+      },
+    });
+    expect(result).toMatchObject({
+      canDelete: true,
+      canEdit: true,
+    });
+  });
+
+  it('trial session can not be deleted when calendared but it sill can be edited if in the future', () => {
```
Comment (waldoj) [inline web-client/src/presenter/computeds/formattedTrialSessionDetails.test.js]:
```suggestion
  it('trial session can not be deleted when calendared but it still can be edited if in the future', () => {
```

File: web-client/src/presenter/computeds/orderTypesHelper.test.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      { code: 'Shenzi', documentType: 'Hyena', eventCode: 'O' },
+    ],
+    USER_ROLES: {
+      petitionsClerk: 'petitionsclerk',
+    },
+  },
+};
+
+describe('orderTypesHelper', () => {
+  it('should filter out and only return type O for docketclerk', () => {
```
Comment (waldoj) [inline web-client/src/presenter/computeds/orderTypesHelper.test.js]: This test description isn't right.

File: web-client/src/presenter/computeds/scanBatchPreviewerHelper.test.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    it('returns correct values when no scanner is selected', () => {
+      let testState = { ...state };
+
+      const result = runCompute(scanBatchPreviewerHelper, {
+        state: testState,
+      });
+      expect(result.scannerSourceDisplayName).toEqual('None');
+    });
+
+    it('returns correct values when a scanner is selected and is using single sided', () => {
```
Comment (waldoj) [inline web-client/src/presenter/computeds/scanBatchPreviewerHelper.test.js]: Where here is it being established that it's single-sided?

File: README.md [hunk 7]
```diff
@@ -57,7 +57,7 @@ To exercise the CI/CD pipeline locally, run the following:
 
 This will run the linter, Shellcheck, audit, build, test, Cypress, Cerebral tests, Pa11y, etc. over all the components.
 
-## Running / verifing the project via Docker
+## Running / verifying the project via Docker
```
[the remark this answers is not in the document]
Comment (ericsorenson) [reply README.md]: Updated!

File: README.md [hunk 8]
```diff
@@ -71,6 +71,9 @@ Assuming you have Docker installed, the following command will spin up a Docker
 - You can access S3 local at http://localhost:9000
 - You can access the style guide at http://localhost:1234/style-guide
 
+### ECR
```
[the remark this answers is not in the document]
Comment (ericsorenson) [reply README.md]: @rachaelparris can you please look into this?

File: README.md [hunk 9]
```diff
@@ -71,6 +71,9 @@ Assuming you have Docker installed, the following command will spin up a Docker
...
+ECR is Amazon’s docker container registry that holds images for `ef-cms` builds on CircleCI. Currently, images can be managed in the AWS ECR console under the `ef-cms-us-east-1`. If you need to update the Docker image, you can do so (with appropriate permissions) by running `./docker-to-ecr.sh`. This command will build an image per the `Dockerfile-CI` config, tag it as `latest` and push it to the repo in ECR.
```
[the remark this answers is not in the document]
Comment (ericsorenson) [reply README.md]: @kkoskelin can you please look into this?

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (rachaelparris) [reply README.md]: This should look correct now, down to 1024px.

File: shared/src/business/useCases/trialSessions/deleteTrialSessionInteractor.test.js (hunk 1 above)
Comment (ericsorenson) [reply shared/src/business/useCases/trialSessions/deleteTrialSessionInteractor.test.js]: The `createISODateString` function is a mocked out current timestamp get. We always get the timestamp from a utility so that we can test easily.

File: shared/src/persistence/dynamsoft/getScannerInterface.test.js (hunk 3 above)
Comment (ericsorenson) [reply shared/src/persistence/dynamsoft/getScannerInterface.test.js]: When the feeder is enabled, the flatbed is not, and when the flatbed is enabled, the feeder is not. (this applies to Amy's scanner, further testing will be needed for the scanners at the court)

File: shared/src/business/useCases/trialSessions/deleteTrialSessionInteractor.test.js (hunk 1 above)
Comment (waldoj) [reply shared/src/business/useCases/trialSessions/deleteTrialSessionInteractor.test.js]: OK, I think it's the title that was confusing me. I read "past now" as "in the future," but it must mean "is in the past." "Past now" could mean either of those two things — it's just a matter of perspective. :) Would you please retitle the test to be more clear?

File: web-client/src/presenter/computeds/orderTypesHelper.test.js (hunk 5 above)
Comment (ericsorenson) [reply web-client/src/presenter/computeds/orderTypesHelper.test.js]: Updated!

File: shared/src/business/useCases/trialSessions/deleteTrialSessionInteractor.test.js (hunk 1 above)
Comment (ericsorenson) [reply shared/src/business/useCases/trialSessions/deleteTrialSessionInteractor.test.js]: Updated!

File: shared/src/persistence/dynamsoft/getScannerInterface.test.js (hunk 3 above)
Comment (waldoj) [reply shared/src/persistence/dynamsoft/getScannerInterface.test.js]: Ah-ha — thanks!

File: web-client/src/presenter/computeds/scanBatchPreviewerHelper.test.js (hunk 6 above)
Comment (ericsorenson) [reply web-client/src/presenter/computeds/scanBatchPreviewerHelper.test.js]: Single sided is the default. The test description has been updated.

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (ericsorenson) [reply README.md]: Updated!

**Comment by waldoj** [comment]
In @JessicaMarine's absence, I have merged this pull request.
